### ISSUE TITLE
[RFC] HiCache stage 5 (partial): device-side Recurrent + SWA

### ISSUE THREAD

**Issue by MokusMokun** [sgl-project/sglang-jax#1338]
## Summary

On top of `UnifiedRadixCache` (Stage 1, #1311), add two device-side `TreeComponent`s: `RecurrentComponent` (KDA / GDN / GLA linear-recurrent models; realizes the #1280 roadmap's `RecurrentStateComponent` checkbox) and `SWAComponent` (sliding-window attention). This lets recurrent state and SWA KV enter the prefix tree and be reused across requests, the same way full KV is.

This is a **partial delivery** of #1280 Stage 5: only the device-side component logic. The HiCache-coupled items (host backup, `--hicache-ratio` benchmark, PD coordination) are **out of scope here** and remain with HiCache (Stage 3/4).

## Position in the roadmap

#1280 Stage 5 ("Multi-component + tuning") mixes two kinds of checkboxes:

| Stage 5 checkbox | True dependency | This issue |
|---|---|---|
| `SWAComponent` | Stage 1 (core) | ✅ device-only |
| `RecurrentStateComponent (Mamba/KDA)` | Stage 1 (core) | ✅ device-only |
| Benchmark across `--hicache-ratio` | Stage 3/4 (HiCache) | ❌ deferred |
| Coordinate with PD on HBM/host split | PD + Stage 3/4 | ❌ deferred |

**Prerequisite = #1311 (Stage 1)** only; no dependency on HiCache (Stage 2/3/4) — HiCache hooks remain no-op placeholders here.

## Sub-issues

| Sub | Scope | Size | Depends on |
|---|---|---|---|
| S5a | Recurrent "make it correct": remove the `--disable-radix-cache` guard + recurrent capability dispatch (`supports_mamba()`) + 3-state slot lifecycle + release-path split (don't free the recurrent snapshot at request finish) + slow-path CoW (full state) + `enable_mamba_extra_buffer` for page≥128 | L | #1311 complete |
| S5b | Recurrent "make it fast": h0-clone Pallas kernel + JIT buffer donation | L | S5a |
| S5c | `SWAComponent` + scheduler SWA capability methods (port upstream `supports_swa()` / `is_chunk_cache()` / `is_tree_cache()`; replace the cache-typed `isinstance` checks) | L | S5a |

Dependency (S1a = typed-params, S1b = core, S1c = M1 tests, all under #1311):

```
#1311: S1a ──► S1b ──► S1c ──► [Stage 1 complete] ──► S5a ──┬──► S5b
                                                            └──► S5c
```

S5a starts after **#1311 complete** (S1a + S1b + S1c green) — build Recurrent on a validated Stage 1, matching the RFC's M1 → M2 → M3 critical path. (Code-wise recurrent only uses the S1a/S1b interfaces, but the milestone/merge condition is "Stage 1 fully tested".) **S5c depends on S5a** — not for recurrent logic (S5c does not use `RecurrentComponent`), but because both extend the same `BasePrefixCache` capability surface (S5a adds `supports_mamba()`; S5c adds `supports_swa()` / `is_tree_cache()` / `is_chunk_cache()`) and both edit `schedule_batch.py` + `common.py`; running them in parallel would race on the same base interface, so S5a establishes the capability pattern once and S5c extends it. This also matches the RFC's recurrent-before-SWA (M2 → M3) priority. Exception: the h0-clone kernel inside S5b is an isolated op (no cache dependency) that may start during Stage 1 in parallel, but its integration into S5b still follows S5a.

## Proposed structure changes

- `sgl_jax/srt/mem_cache/unified_cache_components/recurrent_component.py` — **new**: `RecurrentComponent` (port upstream `MambaComponent`, rename `MAMBA`→`RECURRENT`)
- `sgl_jax/srt/mem_cache/unified_cache_components/swa_component.py` — **new**: `SWAComponent`
- `sgl_jax/srt/mem_cache/memory_pool.py` — **modify**: recurrent slot lifecycle from per-request free to active/tree/free 3-state (`HybridReqToTokenPool`)
- `sgl_jax/srt/mem_cache/common.py` — **modify**: the finish-path `req_to_token_pool.free(req)` currently triggers `HybridReqToTokenPool.free` → `free_recurrent_cache` (frees the slot); split "free the request's token row" from "retain the recurrent snapshot committed to the tree"
- `sgl_jax/srt/model_executor/model_runner_kv_cache_mixin.py` — **modify**: remove the hard guard forcing `--disable-radix-cache` for recurrent; allow recurrent radix only under the unified flag
- `sgl_jax/srt/mem_cache/base_prefix_cache.py` + `sgl_jax/srt/managers/{schedule_batch,schedule_policy}.py` + `sgl_jax/srt/mem_cache/common.py` — **modify (recurrent capability, S5a)**: add `supports_mamba()` and key the hit-path `cow_mamba` (`schedule_batch`), the hybrid-SSM **KV admission** budget (`schedule_policy.rem_total_tokens`), and the recurrent **slot** budget + `release_kv_cache` branch (`common` alloc path) on it instead of the pool-typed `isinstance(HybridReqToTokenPool)`
- **`enable_mamba_extra_buffer` plumbing (page≥128, no such symbol in sgl-jax today, S5a)**: `server_args.py` + `cache_init_params.py` (`CacheInitParams`) + `kv_cache_builder.py` + `recurrent_state_pool.py` / `memory_pool.py` (per-request ping-pong buffer)
- `sgl_jax/srt/kernels/` — **new**: h0-clone Pallas kernel (temporal+conv full state, slot→slot)
- `sgl_jax/srt/managers/{schedule_batch,scheduler}.py` + `sgl_jax/srt/mem_cache/common.py` — **modify**: port upstream's `BasePrefixCache` capability methods (`supports_swa()` / `is_chunk_cache()` / `is_tree_cache()`) and replace the cache-typed `isinstance` checks with them (tree-SWA path guarded by `is_tree_cache() and supports_swa()`, since `SWAChunkCache.supports_swa()` is also True); the allocator-typed checks (`isinstance(allocator, SWATokenToKVPoolAllocator)`) stay

### 1. RecurrentComponent CoW (core)

Recurrent state is overwritten in place (`h_t = f(h_{t-1}, x_t)`), unlike KV which is append-only and read-only-shareable. On a prefix fork, state must be cloned for the new branch, otherwise the write corrupts the original request.

- **Clone scope (correctness-critical)**: a slot's state = temporal/recurrent buffer + conv buffer, one per recurrent layer. Cloning only temporal makes post-hit logits non-equivalent.
- **Lock = single-node-lock**: clone locks only the CoW source node (`best_match_node`), unlike FULL/SWA path-to-root.
- **3-state slot invariant**: `active + tree + free == pool capacity`.

### 2. Two clone paths

| Path | Mechanism | Stage |
|---|---|---|
| Slow path | JAX `dynamic_update_slice` full-buffer copy, no kernel needed | S5a (correct) |
| Fast path | Pallas h0-clone kernel (slot→slot in-place clone of temporal+conv full state) + JIT buffer donation (reuse the input buffer's memory as the output, saving one copy) | S5b (fast) |

Prefix reuse lands correctly in S5a without the kernel; S5b only makes it fast.

### 3. Recurrent under page≥128

With the recommended UHRC test config page≥128, recurrent must use `enable_mamba_extra_buffer` (a per-request ping-pong buffer caching the recurrent state of the partial trailing page) — the largest single piece here, in S5a.

### 4. SWAComponent and scheduler trait-ization

`SWAComponent` reuses the existing `SWARadixCache` logic. But the scheduler currently dispatches SWA behavior by concrete class (SWA allocator pairing, SWA evict); this must become capability (trait) based so `UnifiedRadixCache` can take the SWA path.

## Testing

- Guarantee the slow-path clone is correct: KL/logit equivalence test (hit-path vs no-cache recompute, KL == 0), blocking in **S5a** — this is what proves the clone covers temporal+conv across all layers (catches a temporal-only clone; token-level determinism can mask it).
- Guarantee the fast path matches the slow path: KL/logit equivalence test (fast kernel+donation vs S5a slow path), blocking in **S5b**.
- Guarantee slots don't leak: slot-ledger test (assert `active + tree + free == pool capacity` every step).
- Guarantee hits are reproducible: hit/flush determinism test (cheap token-level oracle).
- Guarantee the kernel beats the slow path: h0-clone kernel microbench (self-compare slow path + HBM-bandwidth roofline).
- Guarantee SWA correctness: SWA KL/logit equivalence (S5c).

## Out of scope (left for full Stage 5 completion)

- recurrent / SWA HiCache host backup (hooks stay no-op here)
- `--hicache-ratio` hit-rate / latency benchmark
- HBM/host memory-split coordination with PD
- `(FULL, SWA, REC)` three components in one tree (upstream structurally unsupported)

**Comment by iwknow** [comment]
i am going to take a look at this issue starting today. please let me know if anyone is working on this.

**Comment by cjx0709** [comment]
@iwknow hello，about S5b, you can see https://github.com/sgl-project/sglang-jax/pull/1509 this pr; and for S5c, @Infra-lab-dev are preparing for the desgin and RFC, if you're interested, you can help with the review and the division of tasks.
